### PR TITLE
create grad_var when run Backward pass

### PR DESCRIPTION
**Pull request by jacquesqiao**
project: https://github.com/PaddlePaddle/Paddle/issues/4679
fix: https://github.com/PaddlePaddle/Paddle/issues/4767

### CODE REVIEW

File: paddle/framework/backward.cc [hunk 1]
```diff
@@ -375,10 +376,18 @@ std::vector<std::unique_ptr<OpDescBind>> MakeBlockBackward(
     backward_descs.insert(backward_descs.begin() + p.first + 1,
                           std::move(p.second));
   }
+
+  // insert fill one op for target
+  std::unique_ptr<OpDescBind> fill_one_op(new OpDescBind(
```
Comment (reyoung) [inline paddle/framework/backward.cc]: It should be in `AppendBackward`. That logic is wrong.

File: paddle/framework/backward.cc [hunk 2]
```diff
@@ -387,11 +396,33 @@ void AppendBackward(ProgramDescBind& program_desc,
     no_grad_var_names.insert(GradVarName(name));
   }
   const int root_block_idx = 0;
-  auto backward_op_descs =
-      MakeBlockBackward(program_desc, root_block_idx, no_grad_var_names);
-  auto& forw_op_descs = program_desc.Block(root_block_idx)->ops_;
+  auto backward_op_descs = MakeBlockBackward(program_desc, root_block_idx,
```
Comment (reyoung) [inline paddle/framework/backward.cc]: target does not need to pass to `MakeBlockBackward`.

File: paddle/framework/backward.cc [hunk 3]
```diff
@@ -387,11 +396,33 @@ void AppendBackward(ProgramDescBind& program_desc,
...
+                                             target, no_grad_var_names);
+  auto root_block = program_desc.Block(root_block_idx);
+  auto& forward_op_descs = root_block->ops_;
```
Comment (reyoung) [inline paddle/framework/backward.cc]: `forward_op_descs` --> `ops`

File: paddle/framework/backward.cc [hunk 4]
```diff
@@ -387,11 +396,33 @@ void AppendBackward(ProgramDescBind& program_desc,
...
   for (auto& ptr : backward_op_descs) {
-    forw_op_descs.push_back(std::move(ptr));
+    forward_op_descs.push_back(std::move(ptr));
+  }
+
+  // create grad_vars in current block.
+  for (size_t index = backward_start_index; index < forward_op_descs.size();
+       ++index) {
+    auto& op = forward_op_descs.at(index);
+    for (auto& input : op->Inputs()) {
```
Comment (reyoung) [inline paddle/framework/backward.cc]: Input does not need to be created

File: paddle/framework/backward.cc [hunk 5]
```diff
@@ -387,11 +396,33 @@ void AppendBackward(ProgramDescBind& program_desc,
     no_grad_var_names.insert(GradVarName(name));
   }
   const int root_block_idx = 0;
-  auto backward_op_descs =
-      MakeBlockBackward(program_desc, root_block_idx, no_grad_var_names);
-  auto& forw_op_descs = program_desc.Block(root_block_idx)->ops_;
+  auto backward_op_descs = MakeBlockBackward(program_desc, root_block_idx,
+                                             target, no_grad_var_names);
+  auto root_block = program_desc.Block(root_block_idx);
```
Comment (reyoung) [inline paddle/framework/backward.cc]: Not only root_block need create vars.

File: paddle/framework/backward.cc (hunk 1 above)
Comment (jacquesqiao) [reply paddle/framework/backward.cc]: done

File: paddle/framework/backward.cc (hunk 2 above)
Comment (jacquesqiao) [reply paddle/framework/backward.cc]: removed

File: paddle/framework/backward.cc (hunk 3 above)
Comment (jacquesqiao) [reply paddle/framework/backward.cc]: done

File: paddle/framework/backward.cc (hunk 4 above)
Comment (jacquesqiao) [reply paddle/framework/backward.cc]: done

File: paddle/framework/backward.cc (hunk 5 above)
Comment (jacquesqiao) [reply paddle/framework/backward.cc]: done

File: paddle/framework/backward.cc [hunk 6]
```diff
@@ -380,24 +381,56 @@ std::vector<std::unique_ptr<OpDescBind>> MakeBlockBackward(
...
                                              &no_grad_var_names, &grad_to_var);
-  auto& forw_op_descs = program_desc.Block(root_block_idx)->ops_;
   for (auto& ptr : backward_op_descs) {
-    forw_op_descs.push_back(std::move(ptr));
+    all_ops.push_back(std::move(ptr));
+  }
+  root_block->NewVar(fill_one_op_out);
+
+  // create grad_var for all blocks in this program
+  for (size_t block_index = 0; block_index < program_desc.Size();
```
Comment (reyoung) [inline paddle/framework/backward.cc]: Not all block is generated by backward.

Comment (jacquesqiao) [reply paddle/framework/backward.cc]: done

**Pull request merged**